### DISCUSSION TITLE
Update on the Homography - Keypoints Model

### DISCUSSION THREAD

**Discussion by DonsetPG** [DonsetPG/narya#20] [Ideas]
Hey,

It seems that at the moment, the biggest issue comes from homography estimation.
Whether it's because the model is not good enough, because images are very different from the dataset we used, or because the model is too slow: there is room for improvement. 

I think that the State of the art right now is, in fact, an evolution of the model Narya uses currently: [A Robust and Efficient Framework for Sports-Field Registration](https://assets.amazon.science/2c/75/0f3bbae44ac7aed02c700bed0694/a-robust-and-efficient-framework-for-sports-field-registration.pdf). 

The model also detects keypoints and dense features, but in a different way.

**Keypoints:**
* Grid keypoints instead of "local" ones 
* Higher weights on the keypoints during training 

**Previous Frame:**
The model uses the previously estimated homography to help the current one. 

**Dense features:**
Instead of directly predicting the homography, the model tries to predict dense features (lines, numbers, etc.). 

However, before building this new model, I think that extending the dataset would be the best thing to do. 

-> If anyone has images that can be used to extend the dataset, please feel free to share them!
-> If anyone wants to help build this new dataset & model, any help would be super appreciated!

Cheers

**Comment by kkoripl** [comment]
Tough thing to be done, especially labeling dataset with keypoints, as any kind of a zoom during transmision changes everything in case of positioning of points placed not on the lines crosses. I think, it's almost impossible to be hand made without any kind of automatic help, even to draw some grid on the pitch just to point where is the region (not exact place) to keypoint to be annotated. Would think about that for my generator.

I have few pictures on [Google Disc](https://drive.google.com/drive/folders/1Q3nbm6TKZXERboqNZessGDAbCXb582lo?usp=sharing) - they are specially made for my A-League stuff, but maybe you'd find something for your usage.

**Comment by karlosos** [comment]
> If anyone wants to help build this new dataset & model, any help would be super appreciated!

I am eager to help with this! Currently, I am doing research on homography estimation (camera calibration) as it is the topic of my thesis. However, I've just started and have nothing more than already implemented in Narya. Feel free to contact me via email.

**Comment by kkoripl** [comment]
Hi, @DonsetPG, when trying to make some code for generating grid keypoints, I came to some conclusions, maybe obvious ones, but:
- to make homography well model need to find at least (in sum) 3 keypoints in two lines - vertical and horizontal;
- which means, a lot of times we're only with points in one line i.e. 16 yard box or halfway line, so as you said a few of new keypoints are needed: what about penalty area, middle of a circle at 16yard box and points on both right and left side in  middle circle;
- I don't know how exactly your model is trained, but what bothers me is 1 pixel bounding box of keypoint. Shouldn't we have bigger one like circle of some, small radius? It's tough to learn and even click on particular keypoint when creating dataset, that makes whole learning process more complex and harder. One more, what about gauss function for loss in the neabourhood of keypoint as it would be strange to harm network if it misses for like 1 pixel from keypoint.

**Comment by DonsetPG** [comment]
Hey,

> to make homography well model need to find at least (in sum) 3 keypoints in two lines - vertical and horizontal;

If using key points, you need at least 4 non-collinear points. 

> which means, a lot of times we're only with points in one line i.e. 16 yard box or halfway line, so as you said a few of new keypoints are needed: what about penalty area, middle of a circle at 16yard box and points on both right and left side in middle circle;

Yes I agree. I think that's why key points based on a grid might be a good approach. 

Regarding what you said before (>  I think, it's almost impossible to be hand made), I think one way to go would be to use the homographic we know for sure (from the dataset), and use them to warp the grid over the image. By doing this, we ensure that the each grid key point is well placed. 

> I don't know how exactly your model is trained, but what bothers me is 1 pixel bounding box of keypoint. Shouldn't we have bigger one like circle of some, small radius? It's tough to learn and even click on particular keypoint when creating dataset, that makes whole learning process more complex and harder. One more, what about gauss function for loss in the neabourhood of keypoint as it would be strange to harm network if it misses for like 1 pixel from keypoint.

They are not 1 pixel bounding box but 9 pixels (cf https://github.com/DonsetPG/narya/blob/master/narya/utils/masks.py):
```python3
def _add_mask(mask, val, x, y):
    """Takes a mask, and add a new segmentation with the value val, around the (x,y) coordinates
    Arguments:
        mask: np.array, the mask
        val: The value to add to the mask 
        x,y: the coordinates of the segmentation to add
    Returns:
        
    Raises:
        
    """
    dir_x = [0, -1, 1]
    dir_y = [0, -1, 1]
    for d_x in dir_x:
        for d_y in dir_y:
            new_x = min(max(x + d_x, 0), mask.shape[0]-1)
            new_y = min(max(y + d_y, 0), mask.shape[1]-1)
            mask[new_x][new_y] = val
```

I don't know if the number of pixels should be increased. 
However, different weights could be applied if the pixel is within a key point or not. 

I think we should try to see how such grid key points could be generated. 
Regarding the rest of the models (player detection and tracking), I think great improvement are currently being made with Graph Neural Networks. I worked on them for Fluid Mechanics so I'll try to see if we can easily apply them to improve Narya overall.

**Comment by DonsetPG** [comment]
Hey,

You can check 
```python3 
def get_field_coordinates(self, bbox, pred_homo, method):
        """Computes the warped coordinates of a bounding box
        Arguments:
            bbox: np.array of shape (4,).
            pred_homo: np.array of shape (3,3)
            method: string in {'cv','torch'}, to precise how the homography was predicted
        Returns:
            dst: np.array, the warped coordinates of the bouding box
        Raises:
        """
        x_1 = int(bbox[0])
        y_1 = int(bbox[1])
        x_2 = int(bbox[2])
        y_2 = int(bbox[3])
        x = (x_1 + x_2) / 2.0 * self.ratio
        y = max(y_1, y_2) * self.ratio
        if method == "cv":
            pts = np.array([float(x), float(y)])
            dst = warp_point(pts, np.linalg.inv(pred_homo), method="cv")
        else:
            pts = np.array([int(x), int(y)])
            dst = warp_point(pts, np.linalg.inv(pred_homo), method="torch")
        return dst
```

in [homography_estimator.py](https://github.com/DonsetPG/narya/blob/master/narya/tracker/homography_estimator.py).

-> bbox comes from the tracking model 
-> pred_homo comes from the method above in the same file: 

```python3 
def __call__(self, input_img):
        pr_mask = self.keypoints_model(input_img)
        src, dst = _points_from_mask(pr_mask[0])
        if len(src) > 3:
            pred_homo = get_perspective_transform(dst, src)
            method = "cv"
        else:
            corners = self.homo_model(input_img)
            pred_homo = compute_homography(corners)[0]
            method = "torch"
        return pred_homo, method
```

-> method: same as above.

I hope this helps.
